File: ruby/lib/data_structures/doubly_linked_list.rb
# frozen_string_literal: true

module DSA
  module DataStructures
    # Doubly linked list with bidirectional traversal.
    # Time: O(1) for prepend/append/remove_first/remove_last, O(n) for arbitrary access
    # Space: O(n)
    class DoublyLinkedList
      Node = Struct.new(:value, :prev_node, :next_node)

      def initialize
        @head = nil
        @tail = nil
        @size = 0
      end

      def prepend(value)
        node = Node.new(value, nil, @head)
        if @head
          @head.prev_node = node
        else
          @tail = node
        end
        @head = node
        @size += 1
        self
      end

      def append(value)
        node = Node.new(value, @tail, nil)
        if @tail
          @tail.next_node = node
        else
          @head = node
        end
        @tail = node
        @size += 1
        self
      end

      def remove_first
        return nil unless @head

        value = @head.value
        @head = @head.next_node
        if @head
          @head.prev_node = nil
        else
          @tail = nil
        end
        @size -= 1
        value
      end

      def remove_last
        return nil unless @tail

        value = @tail.value
        @tail = @tail.prev_node
        if @tail
          @tail.next_node = nil
        else
          @head = nil
        end
        @size -= 1
        value
      end

      def first
        @head&.value
      end

      def last
        @tail&.value
      end

      def get(index)
        return nil if index < 0 || index >= @size

        if index < @size / 2
          current = @head
          index.times { current = current.next_node }
        else
          current = @tail
          (@size - 1 - index).times { current = current.prev_node }
        end
        current.value
      end

      def index_of(value)
        current = @head
        index = 0
        while current
          return index if current.value == value

          current = current.next_node
          index += 1
        end
        -1
      end

      def contains?(value)
        index_of(value) != -1
      end

      def empty?
        @size.zero?
      end

      def size
        @size
      end

      def clear
        @head = nil
        @tail = nil
        @size = 0
        self
      end

      def to_a
        result = []
        current = @head
        while current
          result << current.value
          current = current.next_node
        end
        result
      end
    end
  end
end

File: elixir/lib/data_structures/doubly_linked_list.ex
defmodule DataStructures.DoublyLinkedList do
  @moduledoc """
  Doubly linked list implementation using two lists.
  Time: O(1) for operations at ends, O(n) for arbitrary access
  Space: O(n)
  """

  defstruct front: [], back: [], count: 0

  @type t :: %__MODULE__{front: list(), back: list(), count: non_neg_integer()}

  @doc "Create an empty list"
  @spec new() :: t()
  def new, do: %__MODULE__{}

  @doc "Create from list"
  @spec from_list(list()) :: t()
  def from_list(list), do: %__MODULE__{front: list, count: length(list)}

  @doc "Add element to the front"
  @spec prepend(t(), any()) :: t()
  def prepend(%__MODULE__{front: front, back: back, count: count}, value) do
    %__MODULE__{front: [value | front], back: back, count: count + 1}
  end

  @doc "Add element to the back"
  @spec append(t(), any()) :: t()
  def append(%__MODULE__{front: front, back: back, count: count}, value) do
    %__MODULE__{front: front, back: [value | back], count: count + 1}
  end

  @doc "Remove first element"
  @spec remove_first(t()) :: {:ok, any(), t()} | {:error, :empty}
  def remove_first(%__MODULE__{front: [], back: [], count: 0}) do
    {:error, :empty}
  end

  def remove_first(%__MODULE__{front: [], back: back, count: count}) do
    remove_first(%__MODULE__{front: Enum.reverse(back), back: [], count: count})
  end

  def remove_first(%__MODULE__{front: [head | tail], back: back, count: count}) do
    {:ok, head, %__MODULE__{front: tail, back: back, count: count - 1}}
  end

  @doc "Remove last element"
  @spec remove_last(t()) :: {:ok, any(), t()} | {:error, :empty}
  def remove_last(%__MODULE__{front: [], back: [], count: 0}) do
    {:error, :empty}
  end

  def remove_last(%__MODULE__{front: front, back: [], count: count}) do
    remove_last(%__MODULE__{front: [], back: Enum.reverse(front), count: count})
  end

  def remove_last(%__MODULE__{front: front, back: [head | tail], count: count}) do
    {:ok, head, %__MODULE__{front: front, back: tail, count: count - 1}}
  end

  @doc "Get first element"
  @spec first(t()) :: {:ok, any()} | {:error, :empty}
  def first(%__MODULE__{front: [], back: []}), do: {:error, :empty}
  def first(%__MODULE__{front: [], back: back}), do: {:ok, List.last(back)}
  def first(%__MODULE__{front: [head | _]}), do: {:ok, head}

  @doc "Get last element"
  @spec last(t()) :: {:ok, any()} | {:error, :empty}
  def last(%__MODULE__{front: [], back: []}), do: {:error, :empty}
  def last(%__MODULE__{front: front, back: []}), do: {:ok, List.last(front)}
  def last(%__MODULE__{back: [head | _]}), do: {:ok, head}

  @doc "Get element at index"
  @spec get(t(), non_neg_integer()) :: {:ok, any()} | {:error, :out_of_bounds}
  def get(%__MODULE__{count: count}, index) when index < 0 or index >= count do
    {:error, :out_of_bounds}
  end

  def get(list, index) do
    {:ok, Enum.at(to_list(list), index)}
  end

  @doc "Find index of first occurrence"
  @spec index_of(t(), any()) :: non_neg_integer() | nil
  def index_of(list, value) do
    Enum.find_index(to_list(list), &(&1 == value))
  end

  @doc "Check if element exists"
  @spec contains?(t(), any()) :: boolean()
  def contains?(list, value), do: value in to_list(list)

  @doc "Check if list is empty"
  @spec empty?(t()) :: boolean()
  def empty?(%__MODULE__{count: 0}), do: true
  def empty?(%__MODULE__{}), do: false

  @doc "Get number of elements"
  @spec size(t()) :: non_neg_integer()
  def size(%__MODULE__{count: count}), do: count

  @doc "Convert to list"
  @spec to_list(t()) :: list()
  def to_list(%__MODULE__{front: front, back: back}) do
    front ++ Enum.reverse(back)
  end
end
